src/ocean/text/xml/Document.d

```diff
@@ -2082,6 +2082,95 @@ public class XmlPath(T)
         }
 }
 
+version ( UnitTest )
+{
+    import ocean.core.Test;
+    import ocean.text.xml.DocPrinter;
+}
+
+unittest
+{
+    auto t = new NamedTest("ocean.text.xml.Document");
+    auto doc = new Document!(char);
+    auto printer = new DocPrinter!(char);
+    mstring buffer;
+
+    auto doc1 = "
+<root>123456789
+  <second>second</second>
+  <third>third</third>
+</root>";
+
+    auto doc2 = "
+<root>12345
+  <one>one</one>
+  <two>two</two>
+</root>";
+
+    auto complex_doc = `
+<VAST version="3.0">
+  <InLine>
+    <AdTitle>VAST 3.0 Instream Test</AdTitle>
+    <Creatives>
+      <Creative id="123456" adId="654321"/>
+    </Creatives>
+  </InLine>
+</VAST>`;
+
+    void generateBasicXML (cstring root_value, cstring node1, cstring value1,
+        cstring node2, cstring value2)
+    {
+        // A single text buffer is used to hold the elements to test that the
+        // elements do not just slice the input but actually copy it to the
+        // elements.
+        Array.copy(buffer, root_value);
+
+        auto root = doc.tree.element(null, "root", buffer);
+
+        Array.copy(buffer, node1);
+        root.element(null, node1, buffer);
+
+        Array.copy(buffer, node2);
+        root.element(null, node2, buffer);
+    }
+
+    // Test basic XML file generation.
+    generateBasicXML("123456789", "second", "second", "third", "third");
+    auto result1 = printer.print(doc);
+    t.test!("==")(result1, doc1);
+
+    // Now do a second document generation that re-uses the same root doc and
+    // check for no memory allocations.
+    testNoAlloc({
+        doc.reset();
+        generateBasicXML("12345", "one", "one", "two", "two");
+    }());
+
+    auto result2 = printer.print(doc);
+    t.test!("==")(result2, doc2);
+
+    // Generate a more complex XML document.
```

Comment:
For readability, maybe split this (and `complex_doc`) into a separate unittest? It's a completely separate test to what comes before, IIUC.

Reply:
Yep, will do.